test/xbgpu/test_beamform.py

```diff
@@ -0,0 +1,114 @@
+
+from katgpucbf import COMPLEX, N_POLS
+from katgpucbf.xbgpu.beamform import BeamformTemplate
+
+
+@numba.njit
+def quant(x):
+    """Round to integer and clamp to [-127, 127]."""
+    return np.fmin(np.fmax(np.round(x), -127), 127)
+
+
+@numba.njit
+def beamform_host(in_: np.ndarray, weights: np.ndarray, delays: np.ndarray, beam_pols: np.ndarray) -> np.ndarray:
+    """Implement the beamforming operation on the CPU.
+
+    The input arrays have the same shape and meaning as in :mod:`katgpucbf.xbgpu.beamform`.
+    """
+    n_frames = in_.shape[0]
+    n_channels = in_.shape[2]
+    n_times = in_.shape[3]
+    n_beams = len(beam_pols)
+    out = np.zeros((n_frames, n_beams, n_channels, n_times, COMPLEX), np.int8)
+    for frame in range(n_frames):
+        for channel in range(n_channels):
+            in_c = in_[frame, :, channel, ..., 0] + np.complex64(1j) * in_[frame, :, channel, ..., 1]
+            for beam in range(n_beams):
+                p = beam_pols[beam]
+                w = weights[:, beam]
+                w = w * np.exp(delays[:, beam] * channel * np.complex64(1j) * np.float32(np.pi))
+                for time in range(n_times):
+                    # .copy() is added because otherwise we get a warning that np.dot is
+                    # more efficient on C-contiguous arrays.
+                    accum = np.dot(in_c[:, time, p].copy(), w)
+                    out[frame, beam, channel, time, 0] = quant(accum.real)
+                    out[frame, beam, channel, time, 1] = quant(accum.imag)
+    return out
+
+
+@pytest.mark.combinations(
+    "n_frames, n_channels, n_times, n_antennas",
+    [1, 5],
+    [1, 128, 200, 1025],
+    [1, 128, 256, 321],
+    [1, 4, 19, 80],
+)
+def test_beamform(
+    context: AbstractContext,
+    command_queue: AbstractCommandQueue,
+    n_frames: int,
+    n_channels: int,
+    n_times: int,
+    n_antennas: int,
+) -> None:
+    """Test :class:`.Beamform` by comparing it to a CPU implementation."""
+    beam_pols = [0, 1, 0, 1, 1]
+    n_beams = len(beam_pols)
+
+    template = BeamformTemplate(context, beam_pols)
+    fn = template.instantiate(command_queue, n_frames, n_antennas, n_channels, n_times)
+    fn.ensure_all_bound()
+    h_in = fn.buffer("in").empty_like()
+    h_out = fn.buffer("out").empty_like()
+    h_weights = fn.buffer("weights").empty_like()
+    h_delays = fn.buffer("delays").empty_like()
+    assert h_in.shape == (n_frames, n_antennas, n_channels, n_times, N_POLS, COMPLEX)
+    assert h_out.shape == (n_frames, n_beams, n_channels, n_times, COMPLEX)
+    assert h_weights.shape == (n_antennas, n_beams)
+    assert h_delays.shape == (n_antennas, n_beams)
+
+    rng = np.random.default_rng(seed=123)
+    h_in[:] = rng.integers(-127, 127, h_in.shape)
+    scale = 2.0 / np.sqrt(n_antennas)
+    h_weights[:] = rng.uniform(-scale, scale, h_weights.shape) + 1j * rng.uniform(-scale, scale, h_weights.shape)
+    h_delays[:] = rng.uniform(-0.1, 0.1, h_delays.shape)
```

Comment:
The scale I fiddled with a bit to ensure that some but not all values saturate. The delays are pretty arbitrary but my gut feel is that they're reasonably representative: the delays in the beamformer are expected to be small since the big delays are taken out by the F-engines and these are just for steering within the field of view.

Reply:
Unfortunately I can't really argue with the logic you've outlined there. *However*, I could request you paste the above explainer in `test_beamform`'s docstring - or at least the second half of it. I understand the counter-response would be around
> Well, if we have to explain the thinking here for deliberately chosen values, then we should do it everywhere.

And you might very well be correct, just that we're here now and this seems a bit more out-there to me.

Feel free to ignore, I'm going to keep a note of this explainer either way.